File: Algorithms/202. Happy Number/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        map={}
        while n!=1:
            if map.get(n) is None:
                map[n]=1
                string=str(n)
                sum=0
                for s in string:
                    i=int(s)
                    sum+=(i*i)
                n=sum
            else:
                return False
        return True

File: Algorithms/202. Happy Number/Solution.java
public class Solution {
    public boolean isHappy(int n) {
        String s=String.valueOf(n);
        Map<String, String> map=new HashMap<String, String>();
        map.put(s, "存在");
        int sum=0;
        while (true) {
            for (int i = 0; i < s.length(); i++) {
                int k=Integer.parseInt(s.substring(i,i+1));
                sum+=k*k;
            }
            if (sum==1) {
                return true;
            }
            if (map.get(String.valueOf(sum))!=null) {
                return false;
            }else {
                s=String.valueOf(sum);
                map.put(s, "存在");
                sum=0;
            }
            
        }
    }
}